README.md
- [Parallel Wordle Solver](#parallel-wordle-solver)
    - [Full Documentation](#full-documentation)
    - [Build](#build)
    - [Help](#help)
    - [Default Behavior (no args)](#default-behavior-no-args)
    - [Play 1 game at a time (`-w`)](#play-1-game-at-a-time--w)
    - [Customizing Heuristics (`-m`)](#customizing-heuristics--m)
    - [Entropy Parallelization Params (`-p`, `-f`, `-e`)](#entropy-parallelization-params--p--f--e)
    - [Interactive Mode (`-i`)](#interactive-mode--i)

# Parallel Wordle Solver

### Full Documentation
Please visit [this pdf](parallel_wordle-report.pdf).

### Build

Make sure to `build` the project before you run any of the commands below:
```bash
ᐅ stack build
```

### Help

```bash
ᐅ stack exec wordleSolver-exe -- -h True +RTS -ls -N5
Usage: wordleSolver [OPTION...]
  -i Bool    --interactive=Bool  Interactive mode
  -m String  --method=String     Heuristic method
  -p String  --parMethod=String  Parallel method
  -f Int     --filterChunk=Int   Filter chunk size
  -e Int     --entropyChunk=Int  Entropy chunk size
  -w String  --word=String       Guess word
  -h         --help              Show help
```

### Default Behavior (no args)

- Initiate 2,309 Wordle games and print out the sequence of turns it took to reach the final answer
- Utilizes the entropy heuristics
- Utilizes the "parMap" parallelization strategy
  - Utilizes `filter_chunk_sz = 50` and `entropy_chunk_sz = 30`
```bash
ᐅ stack exec wordleSolver-exe -- +RTS -ls -N8
"cigar"[3]: ["soare","lucid","cigar"]
"rebut"[3]: ["soare","putid","rebut"]
"sissy"[3]: ["soare","flint","sissy"]
...
```

### Play 1 game at a time (`-w`)

Instead of playing all 2,309 games, we can choose to play only one game by specifying the `-w` option. For example, below is an example of playing one round of Wordle with the answer "baker"
```bash
ᐅ stack exec wordleSolver-exe -- -w "baker" +RTS -ls -N5
["soare","clipt","gombo","deked","baker"]
```

### Customizing Heuristics (`-m`)

While the default heuristics is Entropy, we can choose to use other heuristics. Here are three ways of solving the `prove` wordle game.
```bash
ᐅ stack exec wordleSolver-exe -- -m "frequency" -w "prove" +RTS -ls -N5
["arose","drone","trope","probe","prove"]
ᐅ stack exec wordleSolver-exe -- -m "minimax" -w "prove" +RTS -ls -N5  
["raise","drone","probe","prove"]
ᐅ stack exec wordleSolver-exe -- -m "entropy" -w "prove" +RTS -ls -N5  
["soare","pwned","prove"]
```

### Entropy Parallelization Params (`-p`, `-f`, `-e`)

We can further customize the Entropy Heuristics for it to use different parallelization strategies:
- `-p` specifies the parallelization strategy. Users can choose between `parList` and `parListChunk`
- `-f` specifies the filter chunk size. It's an integer.
- `-e` specifies the entropy chunk size. It's an integer.
Recall that by default, if a user specifies none of the above, we automatically execute `-p parList -f 50 -e 30`.
Here is an example of using the `parListChunk` strategy with filter size of 64 and entropy size of 64:
```bash
ᐅ stack exec wordleSolver-exe -- -m "entropy" -p parListChunk -f 50 -e 50 -w "prove" +RTS -ls -N5  
["soare","pwned","prove"]

ᐅ stack exec wordleSolver-exe -- -m "entropy" -p parListChunk -f 50 -e 50  +RTS -ls -N5  
"cigar"[3]: ["soare","lucid","cigar"]
"rebut"[3]: ["soare","putid","rebut"]
"sissy"[3]: ["soare","flint","sissy"]
...
```

### Interactive Mode (`-i`)

```bash
ᐅ stack exec wordleSolver-exe -- -i True +RTS -ls -N5
Please enter "soare" into NYT's Wordle game.
Enter NYT's response in 0/1/2 format:
...
```

CHANGELOG.md
# Changelog for `wordleSolver`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


app/Main.hs
module Main (main) where

import Parser
import Player
import Interactive ( interactiveLoop )

import System.Environment (getArgs)
import Control.Parallel.Strategies (using, parList, rdeepseq)
import System.Console.ParseArgs ()
import System.Console.GetOpt
    ( getOpt,
      ArgDescr(NoArg, ReqArg),
      ArgOrder(Permute),
      OptDescr(..) )
import Text.Read (readMaybe)

main :: IO ()
main = do
  args <- getArgs
  opts <- wordleOpts args

  valid_words <- fileToWordList "data/answerlist.txt" 
  all_words <- fileToWordList "data/wordlist.txt"

  case opts of
    (Flags {help = True}, _) -> putStrLn usageMsg
    (Flags {interactive = True}, _) -> interactiveLoop valid_words all_words
    (Flags {word = "", method = m, filter_chunk_sz = f_sz, 
            entropy_chunk_sz = e_sz, par_method = pm}, _) -> do
      let arrEntropy = Prelude.map (\w ->
            playAndGetList m w valid_words all_words f_sz e_sz pm)
            valid_words `using` parList rdeepseq
      printGameRes arrEntropy valid_words
    (Flags {word = w, method = m, filter_chunk_sz = f_sz, 
            entropy_chunk_sz = e_sz, par_method = pm}, _) -> do
      let answer = playAndGetList m w valid_words all_words f_sz e_sz pm
      print answer
      -- uncomment the following when doing performance testing
      -- _ <- playAndGetList m w valid_words all_words f_sz e_sz pm `seq` return ()
      -- return ()

data Flags = Flags
  {
    interactive :: Bool,
    method :: String,
    par_method :: String,
    filter_chunk_sz :: Int,
    entropy_chunk_sz :: Int,
    word :: String,
    help :: Bool
  } deriving Show

defaultFlags :: Flags
defaultFlags = Flags
  {
    interactive = False,
    method = "entropy",
    par_method = "parMap",
    filter_chunk_sz = 50,
    entropy_chunk_sz = 30,
    word = "",
    help = False
  }

flags :: [OptDescr (Flags -> Flags)]
flags =
  [ Option ['i'] ["interactive"] (ReqArg (\f opts -> 
      opts { interactive = readInteractive f }) "Bool") "Interactive mode"
  , Option ['m'] ["method"] (ReqArg (\f opts -> 
      opts { method = readMethod f }) "String") "Heuristic method"
  , Option ['p'] ["parMethod"] (ReqArg (\f opts -> 
      opts { par_method = readParMethod f }) "String") "Parallel method"
  , Option ['f'] ["filterChunk"] (ReqArg (\f opts -> 
      opts { filter_chunk_sz = readFilterChunkSz f }) "Int") "Filter chunk size"
  , Option ['e'] ["entropyChunk"] (ReqArg (\f opts -> 
      opts { entropy_chunk_sz = readEntropyChunkSz f }) "Int") "Entropy chunk size"
  , Option ['w'] ["word"] (ReqArg (\f opts -> 
      opts { word = f }) "String") "Guess word"
  , Option ['h'] ["help"] (NoArg (\opts -> 
      opts { help = True })) "Show help"
  ]

readInteractive :: String -> Bool
readInteractive s = case readMaybe s of
  Just b -> b
  Nothing -> error "ERROR: unrecognized option -i [Bool]"

readMethod :: String -> String
readMethod s = case s of
  "entropy" -> "entropy"
  "frequency" -> "frequency"
  "minimax" -> "minimax"
  _ -> error "ERROR: unrecognized option -m [entropy | frequency | minimax]"

readParMethod :: String -> String
readParMethod s = case s of
  "seq" -> "seq"
  "parMap" -> "parMap"
  "parListChunk" -> "parListChunk"
  _ -> error "ERROR: unrecognized option -p [seq | parMap | parListChunk]"

readFilterChunkSz :: String -> Int
readFilterChunkSz s = case readMaybe s of
  Just i -> i
  Nothing -> error "ERROR: unrecognized option -f [Int]"

readEntropyChunkSz :: String -> Int
readEntropyChunkSz s = case readMaybe s of
  Just i -> i
  Nothing -> error "ERROR: unrecognized option -e [Int]"

wordleOpts :: [String] -> IO (Flags, [String])
wordleOpts args =
  case getOpt Permute flags args of
    (xs, n, []) -> return (foldl (flip id) defaultFlags xs, n)
    (_, _, errs) -> ioError (userError (concat errs ++ usageMsg))

usageMsg :: [Char]
usageMsg =
  "Usage: wordleSolver [OPTION...]" ++ "\n" ++
  "  -i Bool    --interactive=Bool  Interactive mode" ++ "\n" ++
  "  -m String  --method=String     Heuristic method" ++ "\n" ++
  "  -p String  --parMethod=String  Parallel method" ++ "\n" ++
  "  -f Int     --filterChunk=Int   Filter chunk size" ++ "\n" ++
  "  -e Int     --entropyChunk=Int  Entropy chunk size" ++ "\n" ++
  "  -w String  --word=String       Guess word" ++ "\n" ++
  "  -h         --help              Show help"